client/cmd/dexc-desktop/app_darwin.go

```diff
@@ -0,0 +1,578 @@
+		windows := cocoa.NSApp().OrderedWindows()
+		len := windows.Count()
+		if len < uint64(maxOpenWindows) {
+			createNewWebView()
+		} else {
+			// Show the last window if maxOpenWindows has been exceeded.
+			winObj := windows.ObjectAtIndex(len - 1)
+			win := cocoa.NSWindow_fromRef(winObj)
+			win.MakeMainWindow()
+		}
+	})
+
+	// MacOS will always send the "applicationShouldTerminate" event when an
+	// application is about to exit so we should use this opportunity to
+	// cleanup. See:
+	// https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428642-applicationshouldterminate?language=objc
+	addMethodToDelegate("applicationShouldTerminate:", func(s objc.Object) core.NSUInteger {
+		cancel()                 // no-op with clean rpc/web server setup
+		wg.Wait()                // no-op with clean setup and shutdown
+		shutdownCloser.Done(log) // execute defer statements
+		shutdownCloser.Success()
+		return core.NSUInteger(1)
+	})
+
+	// "applicationDockMenu" method returns the app's dock menu. See:
+	// https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428564-applicationdockmenu?language=objc
+	addMethodToDelegate("applicationDockMenu:", func(_ objc.Object) objc.Object {
+		// Menu Items
+		newWindowMenuItem := cocoa.NSMenuItem_Init("New Window", objc.Sel(selNewWindow), "n")
+		logsMenuItem := cocoa.NSMenuItem_Init("Open Logs", objc.Sel(selOpenLogs), "l")
+
+		menu := cocoa.NSMenu_New()
+		menu.AddItem(newWindowMenuItem)
+		menu.AddItem(logsMenuItem)
+
+		return menu
+	})
+
+	// MacOS will always send the "windowWillClose" event when an application
+	// window is closing. See:
+	// https://developer.apple.com/documentation/appkit/nswindowdelegate/1419605-windowwillclose?language=objc
+	var noteSent bool
+	addMethodToDelegate("windowWillClose:", func(_ objc.Object) {
+		windowsOpen := nOpenWindows.Add(-1)
+		if windowsOpen > 0 {
+			return // nothing to do
+		}
+
+		err := clientCore.Logout()
+		if err == nil {
+			return // nothing to do
+		}
+
+		if !errors.Is(err, dexCore.ActiveOrdersLogoutErr) {
+			log.Errorf("Core logout error: %v", err)
+			return
+		}
+
+		if !noteSent && cocoa.NSApp().IsRunning() { // last window has been closed but app is still running
+			noteSent = true
+			sendDesktopNotification("DEX client still running", "DEX client is still resolving active DEX orders")
+		}
+	})
+
+	// MacOS will always send this event when an application icon on the dock is
+	// clicked or a new process is about to start, so we hijack the action and
+	// create new windows if all windows have been closed. See:
+	// https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428638-applicationshouldhandlereopen?language=objc
+	addMethodToDelegate("applicationShouldHandleReopen:hasVisibleWindows:", func(_ objc.Object) bool {
+		if !hasOpenWindows() {
+			// dexc-desktop is already running but there are no windows open so
+			// we should create a new window.
+			createNewWebView()
+		}
```

Comment:
[From the docs](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428638-applicationshouldhandlereopen?language=objc):
> The application delegate will also get a chance to respond to the normal untitled document delegate methods. If you implement this method in your application delegate, it will be called before any of the default behavior happens. If you return YES, then NSApplication will proceed as normal. If you return NO, then NSApplication will do nothing. So, you can either implement this method with a version that does nothing, and return NO if you do not want anything to happen at all (not recommended), or you can implement this method, handle the event yourself in some custom way, and return NO.
Miniaturized windows, windows in the Dock, are considered visible by this method, and cause flag to return YES, despite the fact that miniaturized windows return NO when sent an [visible](https://developer.apple.com/documentation/appkit/nswindow/1419132-visible?language=objc) message.


Falling through was intentional. Our only concern is to open a window if non exists.